### DISCUSSION TITLE
Project feedback and handy utilities

### DISCUSSION THREAD

**Discussion by wojtekmaj** [pmndrs/jotai#2745] [Show and tell]
Hey! I've recently migrated from Recoil to Jotai, and wanted to share some feedback as well as utilities I've made along the way to help me out with the process.

### `atomFamilyWithDefault`

A really common pattern in my app was to asynchronously fetch an item from remote store, but also be able to write to it. While for `atom` we have `atomWithDefault`, for `atomFamily` - you're pretty much on your own. So I came up with this:

```ts
export function atomFamilyWithDefault<T, P>(
  getRead: (param: P) => Read<Promise<T>>,
): AtomFamily<P, WritableAtom<T | Promise<T>, [SetStateAction<T | Promise<T>>], void>> {
  return atomFamily(
    (param: P) => atomWithDefault<T | Promise<T>>(getRead(param)),
    dequal,
  ) as AtomFamily<P, WritableAtom<T | Promise<T>, [SetStateAction<T | Promise<T>>], void>>;
}
```

The type assertions are there to make it possible to write synchronous setters, like `setItem(prevItem => ({ ...prevItem, is_active: true }))`. Otherwise, `prevItem` would need to be awaited, but in practice, you can't really (at least in my app) get a promise there. This problem also affects `atomFamily`, so I also have a type-asserting wrapper for it.

### `createJointFamily`

The requirement here was to have a joint family of atoms for:
- Fetching a list of items (e.g. from `/api/business/:id/items`)
- Fetching a single item (e.g. from `/api/business/:id/items/:itemId`)

with the requirement of both`atomFamily` being writable and connected. Specifically:
- A single item should not be fetched in `itemAtomFamily` if it was already fetched by `itemsAtomFamily`
- Writing a new list of items to `itemsAtomFamily` should update the same items in `itemAtomFamily`
- Writing a new list of items to `itemAtomFamily` should update the same item in `itemAtomsFamily`

Usage:
```ts
const { itemsAtomFamily, itemAtomFamily } = createJointFamily(
    (id) => () => getItems(id),
    (ids) => () => getItem(ids),
);
```

Source:

```ts
export function createJointFamily<
  T extends BaseRecord,
  P1 extends string | undefined = string,
  P2 extends P1 extends string ? [P1, string] : string = P1 extends string ? [P1, string] : string,
>(
  getItemsRead: (id: P1) => Read<Promise<T[]>>,
  getItemRead: (param: P2) => Read<Promise<T>>,
): {
  itemsAtomFamily: AtomFamily<
    P1,
    WritableAtom<T[] | Promise<T[]>, [SetStateAction<T[]>], Promise<void>>
  >;
  itemAtomFamily: AtomFamily<P2, WritableAtom<T | Promise<T>, [SetStateAction<T>], Promise<void>>>;
} {
  const itemMap = new Map<string, T>();

  const defaultItemsAtomFamily = atomFamilyWithDefault((param: P1) => async (get) => {
    const items = await getItemsRead(param)(get);

    for (const item of items) {
      itemMap.set(item._id, item);
    }

    return items;
  });

  const itemsAtomFamily = atomFamily(
    (param: P1) =>
      atom(
        async (get) => {
          const items = await get(defaultItemsAtomFamily(param));

          const itemsInMap = items.map((item) => {
            const itemInMap = itemMap.get(item._id);

            if (!itemInMap) {
              throw new Error(`Item not found in map: ${item._id}`);
            }

            return itemInMap;
          });

          return itemsInMap;
        },
        async (get, set, nextValueOrSetter: SetStateAction<T[]>) => {
          const nextValue =
            typeof nextValueOrSetter === 'function'
              ? nextValueOrSetter(await get(defaultItemsAtomFamily(param)))
              : nextValueOrSetter;

          for (const item of nextValue) {
            itemMap.set(item._id, item);
          }

          set(defaultItemsAtomFamily(param), nextValue);
        },
      ),
    dequal,
  );

  const defaultItemAtomFamily = atomFamilyWithDefault((param: P2) => async (get) => {
    const itemInMap = itemMap.get(param[1]);

    if (itemInMap) {
      return itemInMap;
    }

    const item = await getItemRead(param)(get);

    itemMap.set(item._id, item);

    return item;
  });

  const itemAtomFamily = atomFamily((param: P2) => {
    const itemsParam = (Array.isArray(param) ? param[0] : undefined) as P1;

    return atom(
      async (get) => {
        const item = await get(defaultItemAtomFamily(param));

        const itemInMap = itemMap.get(item._id);

        if (!itemInMap) {
          throw new Error(`Item not found in map: ${item._id}`);
        }

        return itemInMap;
      },
      async (get, set, nextValueOrSetter: SetStateAction<T>) => {
        const nextValue =
          typeof nextValueOrSetter === 'function'
            ? nextValueOrSetter(await get(defaultItemAtomFamily(param)))
            : nextValueOrSetter;

        itemMap.set(nextValue._id, nextValue);

        set(defaultItemAtomFamily(param), nextValue);

        set(itemsAtomFamily(itemsParam), (prevItems) =>
          prevItems.map((item) => (item._id === nextValue._id ? nextValue : item)),
        );
      },
    );
  }, dequal);

  return {
    itemsAtomFamily,
    itemAtomFamily,
  };
}
```

A simpler version, supporting only one param (for cases like a pair of APIs: `/api/businesses` and `/api/businesses/:id`):

```ts
export function createSimpleJointFamily<T extends BaseRecord, P extends string>(
  itemsRead: Read<Promise<T[]>>,
  getItemRead: (param: P) => Read<Promise<T>>,
): {
  itemsAtom: WritableAtom<T[] | Promise<T[]>, [SetStateAction<T[]>], Promise<void>>;
  itemAtomFamily: AtomFamily<P, WritableAtom<T | Promise<T>, [SetStateAction<T>], Promise<void>>>;
} {
  const { itemsAtomFamily, itemAtomFamily } = createJointFamily(
    (_param: undefined) => itemsRead,
    getItemRead,
  );

  const itemsAtom = itemsAtomFamily(undefined);

  return {
    itemsAtom,
    itemAtomFamily,
  };
}

```

Hope this helps anyone, whether it's the maintainers or the users :)

**Comment by dmaskasky** [comment]
This is awesome, thanks for sharing.

How has you experience been migrating?
Other than these utilities above, was there any utility or documentation you found lacking or confusing?
How did you migrate? All at once or little by little (e.g in a hybrid state for awhile).

**Reply by wojtekmaj** [reply]
I went all-in and swapped the whole thing at once.

Given the scale of the whole operation (9 stores, ~1050 connected components) I would say it was incredibly smooth. I was particularly happy about the fact that virtually no changes were necessary outside the store; other than a mass find-and-replace operation really.

I was thrilled to see that my apps became, on average, 71 KB lighter without sacrificing anything at all.

There were certain pain points though. By no means I'm trying to criticize, but since you've asked, here they are:

### Async writable atoms 

I find creating async writable `atom`s (and even more so, `atomFamily`s) painful (e.g. using a `fetch()`). 

`atomWithDefault` is making it slightly easier, but I don't see why this helper is needed in the first place.

`atomWithStorage` _would_ have been sooo helpful, too, if I was able to mix it with async writable atoms. But since I couldn't mix them, I had to write my own kinda-writable-async-atom-with-localstorage-sync from scratch to store current user in sessionStorage.

If writable async atoms are a no-go, I at least wish I was able to mix and match the utils, like `withStorage(withDefault(atom(loadAsyncStuff)))`;

### Loadables

Loadables are lacking a bit when compared to Recoil ones. Particularly, I'm missing a `toPromise()` method, which in the past enabled me to do things like:

```
function ExportButton() {
  const heavyAtomLoadable = useRecoilLoadable(heavyAtom);

  async function onClick() {
    // Heavy export operation
    const data = await heavyAtomLoadable.toPromise();

    // Do something with data
  }

  return (<button onClick={onClick}>Export</button>);
}
```

With Jotai, this pattern seems almost impossible to recreate, and certainly not as easy as on the Recoil example.

### Types

I'm a bit confused about the types for setters. Sometimes, its type is `SetStateAction<T>`, other times `SetStateAction<Promise<T>>`, others `SetStateAction<T | Promise<T>>`... 

Perhaps it's my utilities fault, but I feel like asynchronously initialized atoms should still have synchronous prevValues provided in setters. Writing setters like

```ts
setItem(async (prevItemOrPrevItemPromise) => {
  const prevItem = await prevItemOrPrevItemPromise;
  return { ...prevItem, is_active: true };
})
```
 
feels odd, no matter the source of the default value.

The same applies to the third argument of `WritableAtom` type. Sometimes it's `void`, sometimes it's `Promise<void>`, I'm lost here.

At some times, I was forced to write rather silly types:

```ts
  itemQuery: // biome-ignore lint/suspicious/noExplicitAny: Too complex to type
    | AtomFamily<any, WritableAtom<T | Promise<T>, [SetStateAction<T>], void>> // biome-ignore lint/suspicious/noExplicitAny: Too complex to type
    | AtomFamily<any, WritableAtom<T | Promise<T>, [SetStateAction<T>], Promise<void>>>;
```

which caused me to fall deeper into my own trap and type assertion was unavoidable:

```ts
  const setItem = useSetAtom(
    itemQuery(itemId) as WritableAtom<T | T[], [SetStateAction<T | T[]>], void>,
  );
```

otherwise `setItem` type was `never`.

### `atomFamily` and `dequal` trap 

While I understand the rationale for a different approach of comparing params, I forgot to add `dequal` as a second argument of `atomFamily`, and the experience of finding it out wasn't great. The only symptom was 100% CPU usage, and it took me a while to track down the atom family at fault.

**Reply by dmaskasky** [reply]
This is very helpful, thank you.
CC: @dai-shi

**Reply by dai-shi** [reply]
> If writable async atoms are a no-go, I at least wish I was able to mix and match the utils, like `withStorage(withDefault(atom(loadAsyncStuff)))`;

Yeah, this is one of the big limitations. I hope `jotai-effect` will make `withStorage` possible in the future. (I have a task in my action items.)

**Reply by dai-shi** [reply]
> Types

Can you provide a small example with https://www.typescriptlang.org/play ?
I think it requires a few techniques to make types happy.

**Reply by dai-shi** [reply]
> atomFamily and dequal trap

That's hard to solve from the library perspective. One of the pitfalls is that we name it `atomFamily` which has a different capability and a different mental model from Recoil's `atomFamily`.
I would create a wrapper that requires the second argument on the app end.

**Reply by wojtekmaj** [reply]
> Can you provide a small example with https://www.typescriptlang.org/play ?
I think it requires a few techniques to make types happy.

I did my best to illustrate the problem! Hope this helps.

[TS Playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgQxhEAaOBXAzgUwEE0QA1ZAG23yz3wGV8Zj04BfOAMylYHIArCDGTBeAbgBQoSLEQoSAMWQhgFAJ5ZU6AOrAYACwAi+TsmwV4Hbn0HDgAemwxVucVPDR4SACb4AjtiU7Fw8IHC8vgGU9hR6+G7uMvAAVCi4cABK+MgAxvDWYbxQ2Xlu0p5wMGpg+HIsmHCMMPTCMER5wBAAdljaUHrIAEYURCTBBeG2ImUeslU1dYrKqmrjoZNCIvYAbshdqhTIjs4UuPZaIEoq6gnztQDiTG1QcAC8cAA85FT4AHwAFBcAFxwepfSjUX4ASjevzg32okgkd0y2W84J+cPe-wA5kwQY8YM8Ya84Qj8EiUQBJNphd5IYDeEFdbAgQb4KBiOBdZT4EG4GD9Lo4rnAXAAfVyzm2fLggwgEBGe3YSM42C6HW68nQVxWugMxlM5hgHwAKlgAAoAvEwLLIJlwf5gZBQZQgi0kuF29EW0Ji-Bm37QxASOBwYowbBQLray7LdT-J0ut1wD2w2P6owmMwWM1wAA+qb9BED1qY3qTrpAUKhWEigQoUMkbAkEnsyWSodSpv0YrSBFgnRjOXQ+HSwBj+j23g0cvwOTMBDlq18RosWCaLVQ7WcWuKOXwwBl6V96H9gbSKEq1XwADo4F24FSNdB9zB1FgqQBZADy957fbILgA67jGIDIAA1mOsa6uocAEESHLpLgo4AO5TvAqHQBBQwjHAqF6Po94-gYHIEQQn4RI+XT4Pg3iVBAcBTjKaRqBq8FPEhmioSIzjCnAYDFNs5K3l29gSGqGqgTB8ZqJmho5swXTeIQwEcqBuBmpaZa2miIKVimaakqi9ofKeKglqaQYwggobhkwUYxhcsFqP8dlhgZIDup6GaEQpxp5oW5nnlZOkVs6VY1hg7lwPWlDRWGMJAaCSzXGoZm9P0wjDKM6CBUWZ6Wb8WAANqbq0O5DoGAC6WDbBAjJBs2rYjl0ArwaO9BsTk5JvL5BrZgFNL4CAAJ2f8Pn-AyDoAIxYDyICyrw+j4BQFAQLwWBipKHQyiCpinLUbAwrgqBipwwDQcNmASE2LXdO1KGLap3W9e8FzyYNubBSW11Bm5YZAd1jqTdNIIAEzzbyILLRAqEzptcDbVKR6yoK1DsCdZ24BdV20tFd0SK1j2da9ELQe9qUrImjLMqy7JQJ6dkXADYYhmzbO00j3gJRz3LQ+EuDYGAm0xWGyO7WjUDULzYYcKdzg45d6TXbLjo2lgCHzfgAAeMDkj5CBsLzta3UixPwE9RC4GTPzpJTOqyZ9a4mqr3L0xyY0eVzLJshyPlA+xE2wjFU1i9zIKMmrYYLUtQsi9HSMSije2VNLNAxcd8HY7jKv42b91tZbo4vRq5K4JmqkgUO9syWlzuKYQylV+pNes-8Pse4z6aBzkIPpmHHNc1HMWxzDPHAcLotD8nksgujGdy6bhPtp2cDdr26Qjh4NFdPAxSCWO+B7+kpHc9kFBYLrNR5HR141OknDQHAAAGqFZbhuUgC-onr+JkmahjKaCAOIcQjEIJLQe+EP45XqNFNgIIkDvwGLAkgII+goPASQD4bsypMC3G0CBoEcG0l+LVOA9VGqY3ZnAew9hoGYK-r1FEYosAQE4HAEcUYKL9mADiHkOUGJPlpL-ThD1MIwKwegckIJrp9ToPUck-xkHZSkdWSQYYLYcRgBg1RX95EECaPUZRkiv6EzDAA6S3QADCsQcgQWDrZDmCFdGf2MWDOAc1+aLRhitNaG0tqz2lFLDGx0NFszoQJISvU+wjigG+OCE5OAcmKPRZK11RFhhcaY4xTponk1BnAW8xTD7CXJoEnawSQSlJEhLYJmM7py1bGGCMjlHQxQ+IMJwaAYw2LsRBV4CA+nAHsRwO4rwABEXSiTdAmb8cOTQhEp3wB0+w0yenzMSs1NsHZHzPiiSjA8TE4b4BlFAOsX14C7DtlwZ+Fwt6jhCKwXY-QIB4A6lGA8uBNaPPYfhXsOR9AoGKKxDUCVUhW2BbUXw-QZT0QmEIFaLx7n3gUM-M+29IC73gGgFAOQDxgBxStMIlBL74VqDRO+uK6BXlMFAR8IBoC1ExSMHW99ahPxeG-HJJAf5iQkuqQBcBgGgPAZLHsqAIEEpgLgQga0oEqLcSQeBiC7KKtQegIEMVCyuI1SAEhI0CwFQsgGP6pVyrbiIVVORQViymtIWQuqDVvCbLZjqnleUbXGvPGauAeDmgVStd0A1o1yGUJddquAuq1EhvNfgwNgCQ2Oooc611YZ3WMLBD9e1I1ip+otYQxN2ak3JvDb8ZsNk7KRIAHJCFqAokgvUFwxgFAcWKF98KERQGtOAfzSJLjuLgAAhAwvRijybsqTrw-hn8hEZMaIxPQcBrGEGrXOeC4FYBDrslo9VaiZHCMNe8Bt0jyYmMzSQRpYii7aOjfo49himDGL3WYpEFjBVWK6LYkZDjK3OKYHe9xEdPFQx8eEPx61EZ1NRvPdOmNwlhkicU280ykYwF4OkLoQh7w1InX2F+ew1AvywKaG89Acj9EJXAfgeB4CDBEKcXtTgmIcjvE+WK3ReB0cOItNDlRBhAsALwbgBQPZitki96BEy4Z+IU5D0mZZJ0qTBqJpzalBNRg08JLY7KtOjO0jmnTulamGfYwZJmIJjJvJM9Zsy01s0Wbi5ZqybNdFdXdbTlihzCpAWAhg3UrUykcTphyenWZhg+HZ8LIrfMBdqC++ogyrZdTLuTDg9hIufGi2K4JErmD4vwIS2VPb4skES6TFLPw0t2Y+Oluy7nWyea1Fl623VgGxaCy0kLMYwufAy2aHz2WNMlfQGV56NsKvUCq+HfrorKoyly1KgrMq5UUFHUqkbCAralx6ql2h1XatbIkB5j9XnmvJZyLFgAQmoKk3gprc3YIg4DvsGbUKcfZSMoWOl9ea7FtberRt+Ym2ODu3goRTb5plgbc38ALfy4Vlb-21GA-O+XUH4O9urLc9sxrQDofbauzdu701HtyE7n7F4Wd3u6e6996bv3JZI6-oD7baPGQY4O3zGbMXxUYUWwj4rHqQAs-Gztu26OIdsxq9jo7DWTtNfx6Lwnt3K5qUHA9e7jJSceJexyN7wXPu04Mz9-HjPhvC82yXUX5dVfVw1+zyXBmGc5b5-D5bguJMW629b8mFdCIt3V21CXmODOc-q0AA)

**Reply by wojtekmaj** [reply]
> That's hard to solve from the library perspective. One of the pitfalls is that we name it atomFamily which has a different capability and a different mental model from Recoil's atomFamily.
I would create a wrapper that requires the second argument on the app end

I think some kind of a safety net, kinda like the one you're getting if you call React component recursively, would be nice. Like, if you update it 1000 times in a loop, then it bails out with a console error or something.

**Reply by dai-shi** [reply]
> Like, if you update it 1000 times in a loop, then it bails out with a console error or something.

Isn't it what React StrictMode doing already?

**Reply by wojtekmaj** [reply]
You'd think :D Try it!

https://codesandbox.io/p/sandbox/zft2j8

(warning: this will hang your browser)

**Reply by dai-shi** [reply]
It's not visible, but I don't want to hang my browser anyway. If it's async, React can't detect the loop.

**Reply by wojtekmaj** [reply]
Exactly why I believe it would be handy to have this safety net on Jotai side.

**Reply by dai-shi** [reply]
Yeah, I understand it. But, if it can cause without Jotai, it's better for React to fix it.
Is it possible for you to copy files to https://stackblitz.com from codesandbox?

**Comment by dai-shi** [comment]
Re: https://github.com/pmndrs/jotai/discussions/2745#discussioncomment-10729365

> Loadables are lacking a bit when compared to Recoil ones. Particularly, I'm missing a `toPromise()` method, which in the past enabled me to do things like:

Before converting to loadable, it's a promise. However, useAtomValue `use` the promise, so you can't get it. We can avoid it by wrapping it. A workaround is something like this:

```js
const heavyPromiseAtom = atom((get) => ({ promise: get(heavyAtom) }));

function ExportButton() {
  const { promise } = useAtomValue(heavyPromiseAtom);

  async function onClick() {
    // Heavy export operation
    const data = await promise;

    // Do something with data
  }

  return (<button onClick={onClick}>Export</button>);
}
```

**Reply by wojtekmaj** [reply]
Wow, that's really handy! Thank you!

**Comment by dai-shi** [comment]
Re: https://github.com/pmndrs/jotai/discussions/2745#discussioncomment-10751354

> I did my best to illustrate the problem! Hope this helps.
> 
> [TS Playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgQxhEAaOBXAzgUwEE0QA1ZAG23yz3wGV8Zj04BfOAMylYHIArCDGTBeAbgBQoSLEQoSAMWQhgFAJ5ZU6AOrAYACwAi+TsmwV4Hbn0HDgAemwxVucVPDR4SACb4AjtiU7Fw8IHC8vgGU9hR6+G7uMvAAVCi4cABK+MgAxvDWYbxQ2Xlu0p5wMGpg+HIsmHCMMPTCMER5wBAAdljaUHrIAEYURCTBBeG2ImUeslU1dYrKqmrjoZNCIvYAbshdqhTIjs4UuPZaIEoq6gnztQDiTG1QcAC8cAA85FT4AHwAFBcAFxwepfSjUX4ASjevzg32okgkd0y2W84J+cPe-wA5kwQY8YM8Ya84Qj8EiUQBJNphd5IYDeEFdbAgQb4KBiOBdZT4EG4GD9Lo4rnAXAAfVyzm2fLggwgEBGe3YSM42C6HW68nQVxWugMxlM5hgHwAKlgAAoAvEwLLIJlwf5gZBQZQgi0kuF29EW0Ji-Bm37QxASOBwYowbBQLray7LdT-J0ut1wD2w2P6owmMwWM1wAA+qb9BED1qY3qTrpAUKhWEigQoUMkbAkEnsyWSodSpv0YrSBFgnRjOXQ+HSwBj+j23g0cvwOTMBDlq18RosWCaLVQ7WcWuKOXwwBl6V96H9gbSKEq1XwADo4F24FSNdB9zB1FgqQBZADy957fbILgA67jGIDIAA1mOsa6uocAEESHLpLgo4AO5TvAqHQBBQwjHAqF6Po94-gYHIEQQn4RI+XT4Pg3iVBAcBTjKaRqBq8FPEhmioSIzjCnAYDFNs5K3l29gSGqGqgTB8ZqJmho5swXTeIQwEcqBuBmpaZa2miIKVimaakqi9ofKeKglqaQYwggobhkwUYxhcsFqP8dlhgZIDup6GaEQpxp5oW5nnlZOkVs6VY1hg7lwPWlDRWGMJAaCSzXGoZm9P0wjDKM6CBUWZ6Wb8WAANqbq0O5DoGAC6WDbBAjJBs2rYjl0ArwaO9BsTk5JvL5BrZgFNL4CAAJ2f8Pn-AyDoAIxYDyICyrw+j4BQFAQLwWBipKHQyiCpinLUbAwrgqBipwwDQcNmASE2LXdO1KGLap3W9e8FzyYNubBSW11Bm5YZAd1jqTdNIIAEzzbyILLRAqEzptcDbVKR6yoK1DsCdZ24BdV20tFd0SK1j2da9ELQe9qUrImjLMqy7JQJ6dkXADYYhmzbO00j3gJRz3LQ+EuDYGAm0xWGyO7WjUDULzYYcKdzg45d6TXbLjo2lgCHzfgAAeMDkj5CBsLzta3UixPwE9RC4GTPzpJTOqyZ9a4mqr3L0xyY0eVzLJshyPlA+xE2wjFU1i9zIKMmrYYLUtQsi9HSMSije2VNLNAxcd8HY7jKv42b91tZbo4vRq5K4JmqkgUO9syWlzuKYQylV+pNes-8Pse4z6aBzkIPpmHHNc1HMWxzDPHAcLotD8nksgujGdy6bhPtp2cDdr26Qjh4NFdPAxSCWO+B7+kpHc9kFBYLrNR5HR141OknDQHAAAGqFZbhuUgC-onr+JkmahjKaCAOIcQjEIJLQe+EP45XqNFNgIIkDvwGLAkgII+goPASQD4bsypMC3G0CBoEcG0l+LVOA9VGqY3ZnAew9hoGYK-r1FEYosAQE4HAEcUYKL9mADiHkOUGJPlpL-ThD1MIwKwegckIJrp9ToPUck-xkHZSkdWSQYYLYcRgBg1RX95EECaPUZRkiv6EzDAA6S3QADCsQcgQWDrZDmCFdGf2MWDOAc1+aLRhitNaG0tqz2lFLDGx0NFszoQJISvU+wjigG+OCE5OAcmKPRZK11RFhhcaY4xTponk1BnAW8xTD7CXJoEnawSQSlJEhLYJmM7py1bGGCMjlHQxQ+IMJwaAYw2LsRBV4CA+nAHsRwO4rwABEXSiTdAmb8cOTQhEp3wB0+w0yenzMSs1NsHZHzPiiSjA8TE4b4BlFAOsX14C7DtlwZ+Fwt6jhCKwXY-QIB4A6lGA8uBNaPPYfhXsOR9AoGKKxDUCVUhW2BbUXw-QZT0QmEIFaLx7n3gUM-M+29IC73gGgFAOQDxgBxStMIlBL74VqDRO+uK6BXlMFAR8IBoC1ExSMHW99ahPxeG-HJJAf5iQkuqQBcBgGgPAZLHsqAIEEpgLgQga0oEqLcSQeBiC7KKtQegIEMVCyuI1SAEhI0CwFQsgGP6pVyrbiIVVORQViymtIWQuqDVvCbLZjqnleUbXGvPGauAeDmgVStd0A1o1yGUJddquAuq1EhvNfgwNgCQ2Oooc611YZ3WMLBD9e1I1ip+otYQxN2ak3JvDb8ZsNk7KRIAHJCFqAokgvUFwxgFAcWKF98KERQGtOAfzSJLjuLgAAhAwvRijybsqTrw-hn8hEZMaIxPQcBrGEGrXOeC4FYBDrslo9VaiZHCMNe8Bt0jyYmMzSQRpYii7aOjfo49himDGL3WYpEFjBVWK6LYkZDjK3OKYHe9xEdPFQx8eEPx61EZ1NRvPdOmNwlhkicU280ykYwF4OkLoQh7w1InX2F+ew1AvywKaG89Acj9EJXAfgeB4CDBEKcXtTgmIcjvE+WK3ReB0cOItNDlRBhAsALwbgBQPZitki96BEy4Z+IU5D0mZZJ0qTBqJpzalBNRg08JLY7KtOjO0jmnTulamGfYwZJmIJjJvJM9Zsy01s0Wbi5ZqybNdFdXdbTlihzCpAWAhg3UrUykcTphyenWZhg+HZ8LIrfMBdqC++ogyrZdTLuTDg9hIufGi2K4JErmD4vwIS2VPb4skES6TFLPw0t2Y+Oluy7nWyea1Fl623VgGxaCy0kLMYwufAy2aHz2WNMlfQGV56NsKvUCq+HfrorKoyly1KgrMq5UUFHUqkbCAralx6ql2h1XatbIkB5j9XnmvJZyLFgAQmoKk3gprc3YIg4DvsGbUKcfZSMoWOl9ea7FtberRt+Ym2ODu3goRTb5plgbc38ALfy4Vlb-21GA-O+XUH4O9urLc9sxrQDofbauzdu701HtyE7n7F4Wd3u6e6996bv3JZI6-oD7baPGQY4O3zGbMXxUYUWwj4rHqQAs-Gztu26OIdsxq9jo7DWTtNfx6Lwnt3K5qUHA9e7jJSceJexyN7wXPu04Mz9-HjPhvC82yXUX5dVfVw1+zyXBmGc5b5-D5bguJMW629b8mFdCIt3V21CXmODOc-q0AA)

Yeah, it explains.

> ```ts
>     // ...but it's not. prevValue is `any`, TypeScript just bails out here. I don't blame it tbh 😅
>     setWritableAtom((prevValue) => ({ ...prevValue, is_active: prevValue.is_active }));
> ```

Hm, I don't think it's possible. `prevValue` type can't be determined. If TS is clever, it can guess from the return type. I'm not sure if there's anything we can do.

**Reply by wojtekmaj** [reply]
I believe prevValue of a function described as `[SetStateAction<Item>] | [SetStateAction<Item | Promise<Item>>] | [SetStateAction<Promise<Item>>]` should be `Item` or at the very least `Item | Promise<Item>`. I just don't know how we could do it yet. I've asked Matt's TS Wizards on Discord!

**Reply by dai-shi** [reply]
If TS is capable of it, I'm pretty sure we can fix `useSetAtom` typing.

**Reply by wojtekmaj** [reply]
Would that work?

```
type SetStateAction<Value> = Awaited<Value> | ((prev: Awaited<Value>) => Awaited<Value>)
```

See this [TypeScript Playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgQxhEAaOBXAzgUwEE0QA1ZAG23yz3wGV8Zj04BfOAMylYHIArCDGTBeAbgBQoSLEQoSAMWQhgFAJ5ZU6AOrAYACwAi+TsmwV4Hbn0HDgAemwxVucVPDR4SACb4AjtiU7Fw8IHC8vgGU9hR6+G7uMvAAVCi4cABK+MgAxvDWYbxQ2Xlu0p5wMGpg+HIsmHCMMPTCMER5wBAAdljaUHrIAEYURCTBBeG2ImUeslU1dYrKqmrjoZNCIvYAbshdqhTIjs4UuPZaIEoq6gnztQDiTG1QcAC8cAA85FT4AHwAFBcAFxwepfSjUX4ASjevzg32okgkd0y2W84J+cPe-wA5kwQY8YM8Ya84Qj8EiUQBJNphd5IYDeEFdbAgQb4KBiOBdZT4EG4GD9Lo4rnAXAAfVyzm2fLggwgEBGe3YSM42C6HW68nQVxWugMxlM5hgHwAKlgAAoAvEwLLIJlwf5gZBQZQgi0kuF29EW0Ji-Bm37QxASOBwYowbBQLray7LdT-J0ut1wD2w2P6owmMwWM1wAA+qb9BED1qY3qTrpAUKhWEigQoUMkbAkEnsyWSodSpv0YrSBFgnRjOXQ+HSwBj+j23g0cvwOTMBDlq18RosWCaLVQ7WcWuKOXwwBl6V96H9gbSKEq1XwADo4F24FSNdB9zB1FgqQBZADy957fbILgA67jGIDIAA1mOsa6uocAEESHLpLgo4AO5TvAqHQBBQwjHAqF6Po94-gYHIEQQn4RI+XT4Pg3iVBAcBTjKaRqBq8FPEhmioSIzjCnAYDFNs5K3l29gSGqGqgTB8ZqJmho5swXTeIQwEcqBuBmpaZa2miIKVimaakqi9ofKeKglqaQYwggobhkwUYxhcsFqP8dlhgZIDup6GaEQpxp5oW5nnlZOkVs6VY1hg7lwPWlDRWGMJAaCSzXGoZm9P0wjDKM6CBUWZ6Wb8WAANqbq0O5DoGAC6WDbBAjJBs2rYjl0ArwaO9BsTk5JvL5BrZgFNL4CAAJ2f8Pn-AyDoAIxYDyICyrw+j4BQFAQLwWBipKHQyiCpinLUbAwrgqBipwwDQcNmASE2LXdO1KGLap3W9e8FzyYNubBSW11Bm5YZAd1jqTdNIIAEzzbyILLRAqEzptcDbVKR6yoK1DsCdZ24BdV20tFd0SK1j2da9ELQe9qUrImjLMqy7JQJ6dkXADYYhmzbO00j3gJRz3LQ+EuDYGAm0xWGyO7WjUDULzYYcKdzg45d6TXbLjo2lgCHzfgAAeMDkj5CBsLzta3UixPwE9RC4GTPzpJTOqyZ9a4mqr3L0xyY0eVzLJshyPlA+xE2wjFU1i9zIKMmrYYLUtQsi9HSMSije2VNLNAxcd8HY7jKv42b91tZbo4vRq5K4JmqkgUO9syWlzuKYQylV+pNes-8Pse4z6aBzkIPpmHHNc1HMWxzDPHAcLotD8nksgujGdy6bhPtp2cDdr26Qjh4NFdPAxSCWO+B7+kpHc9kFBYLrNR5HR141OknDQHAAAGqFZbhuUgC-onr+JkmahjKaCAOIcQjEIJLQe+EP45XqNFNgIIkDvwGLAkgII+goPASQD4bsypMC3G0CBoEcG0l+LVOA9VGqY3ZnAew9hoGYK-r1FEYosAQE4HAEcUYKL9mADiHkOUGJPlpL-ThD1MIwKwegckIJrp9ToPUck-xkHZSkdWSQYYLYcRgBg1RX95EECaPUZRkiv6EzDAA6S3QADCsQcgQWDrZDmCFdGf2MWDOAc1+aLRhitNaG0tqz2lFLDGx0NFszoQJISvU+wjigG+OCE5OAcmKPRZK11RFhhcaY4xTponk1BnAW8xTD7CXJoEnawSQSlJEhLYJmM7py1bGGCMjlHQxQ+IMJwaAYw2LsRBV4CA+nAHsRwO4rwABEXSiTdAmb8cOTQhEp3wB0+w0yenzMSs1Ns9CACqL4QCLT3igZSHFahn3AjiEZTFkBgBqF0HZ99ajlW3EQqq5IsSgh4nEdEHyCyOjyfgbYIJCDfLaL88mwZjKgt4nRDEkIoStlXo+Z8USUYHiYnDIFHI6xfXgLsO2XBn4XC3qOEIrBdj9AgHgDqUYDy4E1mS9h+Few5H0CgYorENQJVSFbDltRfD9BlPRCYQgVovBJfeBQz8z7b0gLveAaAUA5APGARVK0wiUEvvhWoNE75KroFeUwUBHwgGgLUOVIwdZPKJS8N+OSSA-zEhJdUgC4DANAeAyWPZUAQNVTAXAhA1pQJUW4kg8DEF2VDag9AQIYqFlcTGkAJCRr-J+gGP6pUXmEMASmsIQViwZtIWQuqDVvCbLZgmh1eU5EFsKkWkaxU4B4OaBVN53Q80looWWitYYq2MLBLg7NlUO1-XIZQ8t8a4CJrUWZQtnas34Lbbm9Nnau0Tt+M2GydlIkADkhC1AUSQXqC4YwCgOLFC++FCIoDWnAZlpElx3FwAAQgYXoxR5MbWAWAnwgReElUZMaIxPQcBrGEF3XOeC4FYAvrslo6NaiZHCNTe8I90jyYmIHSQRpYii7aJnfotDhimDGMQ2YpEFjXVWK6LYkZDjt3OKYIR9xEdPFQx8eEPx61EZ1NRvPdOmNwlhkicU280ykYwF4OkLoQh7w1K-X2F+ew1AvywKaG89Acj9DVXAfgeB4CDBEKce9TgmIcjvE+WK3ReCGcOItSTlRBjssALwbgBQPZitk7D6BEwKZ+IUsTfmZZJ0qfxqJQLalBNRg08JLY7KtOjO0jmnTulamGfYwZ6WIJjJvJM9Zsze1s0WUq5Zqz8tdArXdOLlihzupAWAhg3U3kykcfFhyiXWZhg+IVrrHqGvNdqOR+ogyrZdTLuTDg9geufD6164JPrmAqvwGqwNd6hskBG6TcbPxJuFY+FNuyVXWw1a1LN623VgEDday09rMZOufGm2aerc3ovrfQJt56NttvUF2+HJ7nrKoygW365bAag0UHfWG97CAralx6hN2he2DtbIkNV6jtWztjZyANgAQmoKk3gprc3YIgtjvsGbUKcfZSMHWOmPbOwNyHSaPuNe+2ODu3goS-b5jN57gP8DA6Wyt8HTO1Es6x+XDnXPEerMq9sk7QC+dw9x-jwn00SdyE7n7F4WcqcJbu3Tv7DPJai6-izuHkvGTS+R3zf7-XvUYRB8Ltb1aQDm6+-Du2Uvuds323L1Hx30enaVx7lXBPK5qUHA9InjINcePJxySnbWacG+S-TpXJu3tu5hyXD35cI-V2j1bn3yXjfzcd0LsHLvvPZ9h3n8mFdCItyj21b3Mvks26O0AA) and go to line 90.

**Reply by dai-shi** [reply]
It's not correct because `prev` can be a promise.
